inloop/solutions/models.py

```diff
@@ -22,23 +27,59 @@ def hash32(obj):
     return hash_chars[hash(str(obj)) % 32]
 
 
-def get_upload_path(obj, filename):
+def get_upload_path(solution_file, filename):
+    return get_solution_upload_path(solution_file.solution, filename)
+
+
+def get_solution_upload_path(solution, filename):
     """
-    Return an upload file path.
+    Return an upload file path for a solution.
 
     All files related to a specific solution will share a common base directory.
     """
-    s = obj.solution
     return "solutions/{year}/{slug}/{hash}/{id}/{filename}".format_map({
-        "year": s.submission_date.year,
-        "slug": s.task.slug,
+        "year": solution.submission_date.year,
+        "slug": solution.task.slug,
         # another "random" level to avoid too many files per slug directory
-        "hash": hash32(s.author),
-        "id": s.id,
+        "hash": hash32(solution.author),
+        "id": solution.id,
         "filename": filename
     })
 
 
+def create_archive(solution):
+    """
+    Create zip archive of all files associated with a solution.
+    """
+    if solution.archive:
+        return
+    with TemporaryDirectory() as tmpdir:
```

Comment:
Why aren't you using `TemporaryFile()`? The only use of `tmpdir` is the join with a file name. The filename in the database can be chosen when constructing the `SimpleUploadFile`.